internal/experiment/generator.go

```diff
@@ -0,0 +1,152 @@
+	ContainerResourcesSelectors []generation.ContainerResourcesSelector
+	// ReplicaSelectors are the selectors for determining what application resources to scan for desired replica counts.
+	ReplicaSelectors []generation.ReplicaSelector
+	// IncludeApplicationResources is a flag indicating that the application resources should be included in the output.
+	IncludeApplicationResources bool
+	// Default reader to use instead of stdin.
+	DefaultReader io.Reader
+}
+
+// SetDefaultSelectors adds the default selectors to the generator, this requires that the application
+// already be configured on the generator.
+func (g *Generator) SetDefaultSelectors() {
+	// NOTE: This method is completely arbitrary based on what we think the desired output might be.
+	// This bridges the gap between the powerful selection logic that is implemented vs the simple
+	// selection configuration that is actually exposed on the Application.
+
+	// Always add container resource selectors, conditionally with an explicit label selector
+	var crsLabelSelector string
+	if g.Application.Parameters != nil && g.Application.Parameters.ContainerResources != nil {
+		crsLabelSelector = g.Application.Parameters.ContainerResources.LabelSelector
+	}
+	g.ContainerResourcesSelectors = []generation.ContainerResourcesSelector{
+		{
+			GenericSelector: scan.GenericSelector{
+				Group:         "apps|extensions",
+				Kind:          "Deployment",
+				LabelSelector: crsLabelSelector,
+			},
+			Path:               "/spec/template/spec/containers",
+			CreateIfNotPresent: true,
+		},
+		{
+			GenericSelector: scan.GenericSelector{
+				Group:         "apps|extensions",
+				Kind:          "StatefulSet",
+				LabelSelector: crsLabelSelector,
+			},
+			Path:               "/spec/template/spec/containers",
+			CreateIfNotPresent: true,
+		},
+	}
+
+	// Only add replica selectors if the parameter is explicitly configured
+	if g.Application.Parameters != nil && g.Application.Parameters.Replicas != nil {
+		g.ReplicaSelectors = []generation.ReplicaSelector{
+			{
+				GenericSelector: scan.GenericSelector{
+					Group:         "apps|extensions",
+					Kind:          "Deployment",
+					LabelSelector: g.Application.Parameters.Replicas.LabelSelector,
+				},
+				Path:               "/spec/replicas",
+				CreateIfNotPresent: true,
+			},
+			{
+				GenericSelector: scan.GenericSelector{
+					Group:         "apps|extensions",
+					Kind:          "StatefulSet",
+					LabelSelector: g.Application.Parameters.Replicas.LabelSelector,
+				},
+				Path:               "/spec/replicas",
+				CreateIfNotPresent: true,
+			},
+		}
+	}
+}
+
+// Execute the experiment generation pipeline, sending the results to the supplied writer.
+func (g *Generator) Execute(output kio.Writer) error {
+	return kio.Pipeline{
+		Inputs: []kio.Reader{
+			g.Application.Resources,
+		},
+		Filters: []kio.Filter{
+			&konjure.Filter{
+				Depth:         100,
+				DefaultReader: g.DefaultReader,
+				KeepStatus:    true,
```

Comment:
Should we set `KeepStatus: false` if we're going to strip it below?

Reply:
Keeping the status here ensures objects fetched from Kubernetes using `kubectl` retain their status so we can inspect it during the scan. This is independent from striping the status later (which would ensure the experiment itself doesn't have a status field). In the case where `IncludeApplicationResources == true`, then it would also strip the status of the objects fetched from Kube, but only after giving us the opportunity to scan them.